### ISSUE TITLE
Node hashCode() collides almost always

### ISSUE THREAD

**Issue by MorMundHS-MA** [javaparser/javaparser#2857]
Demonstration (in Kotlin):
``` Kotlin
val cu = parser.parse(javaSource).result.get()

val nodes = HashMap<Node, MutableList<Node>>()
Node.PreOrderIterator(cu).forEach { node ->
    if(nodes.containsKey(node)) {
        nodes[node]!!.add(node)
    } else {
        nodes[node] = mutableListOf(node)
    }
}

nodes.values.forEach {
    assert(it.size == 1)
}
```

This will cause collisions for almost any simple source node, like literals. The simplest example is `BooleanLiteralExpr` always returning a `hashCode()` of 0 and 31 for `false` and `true` literals respectively.

Apologies if this is intended behavior and I just missed the relevant documentation. However the HashCodeVisitor class seems to be intended to be collision free from what I can tell.

**Comment by jlerbsc** [comment]
Hi @MorMundHS-MA the code shows that the hashcode method is based on the HashCodeVisitor class is is defined like this : 

    A visitor that calculates a deep hash code for a node by using the hash codes of all its properties, and the hash codes of all its child nodes (by visiting those too.)

So i think this is intended behavior. 
If you have a more specific bug to report feel free to create a new issue.

**Comment by MorMundHS-MA** [comment]
@jlerbsc Apologies for the vague explanation.

The issue is not really related to the HashCodeVisitor itself.

The problem is that for many Node classes the .hashCode() function returns the same value for clearly different object instances.
This simplest example of this would be:
``` java
var a = true;
var b = true;
```
Even though the two `true` literals are a different Node and have different ranges and token ranges, their `hashCode` function *always* evaluates to 31.

So my actual question is: Are the `Node.hashCode()` functions intentionally not using their `range`and `tokenRange`properties for calculation or could I submit a patch to change this?

Currently this breaks anything using a Map/HashMap to store Nodes.

Please let me know if I should still open a new issues or if you can reopen this one.

**Comment by jlerbsc** [comment]
I can not answer this question because I am not experienced enough on JP. @maartenc @ftomassetti @matozoid do you have more explanation on this subject?

**Comment by matozoid** [comment]
@jlerbsc the suggestion that "a" and "b" give the same hashcode would mean that the SimpleName node always has the same hashcode. You could look at the hashcode visitor to see what it does there.

**Comment by jlerbsc** [comment]
Thanks @matozoid  for your tip.

LiteralBooleanExpr hashcode is based on the value and comment.

    public Integer visit(final BooleanLiteralExpr n, final Void arg) {
        return (n.isValue() ? 1 : 0) * 31 + (n.getComment().isPresent() ? n.getComment().get().accept(this, arg) : 0);
    }

For the others nodes you have to consider HashCodeVisitor.

The unit test below prints recursivly class, content and hashcode of each node.

    @Test()
    void test1() throws IOException {

        String src = "public class Foo {\n" +
                "  boolean a = true;\n" +
                "  boolean b = true;\n" +
                "}";

        CompilationUnit cu = StaticJavaParser.parse(src);

        print(cu);
    }

    private void print(Node node) {
        System.out.println("node" + node.getClass().getName() + " " + node.toString() + " " + node.hashCode());
        node.getChildNodes().forEach(child -> print(child));
    }

Result

    node com.github.javaparser.ast.CompilationUnit public class Foo {

    boolean a = true;

    boolean b = true;
    }  -109461522
    node com.github.javaparser.ast.body.ClassOrInterfaceDeclaration public class Foo {

    boolean a = true;

    boolean b = true;
    } 790946670
    node com.github.javaparser.ast.Modifier public  -1539281903
    node com.github.javaparser.ast.expr.SimpleName Foo 2195482
    node com.github.javaparser.ast.body.FieldDeclaration boolean a = true; -939940005
    node com.github.javaparser.ast.body.VariableDeclarator a = true -1721646565
    node com.github.javaparser.ast.type.PrimitiveType boolean -1163918681
    node com.github.javaparser.ast.expr.SimpleName a 3007
    **node com.github.javaparser.ast.expr.BooleanLiteralExpr true 31**
    node com.github.javaparser.ast.body.FieldDeclaration boolean b = true; -939016484
    node com.github.javaparser.ast.body.VariableDeclarator b = true -1721645604
    node com.github.javaparser.ast.type.PrimitiveType boolean -1163918681
    node com.github.javaparser.ast.expr.SimpleName b 3038
    **node com.github.javaparser.ast.expr.BooleanLiteralExpr true 31**

**Comment by matozoid** [comment]
Ah, I see that we're talking about "true" always returning the same hashcode. But that's expected.

**Comment by MysterAitch** [comment]
> So my actual question is: Are the Node.hashCode() functions intentionally not using their `range` and `tokenRange` properties for calculation or could I submit a patch to change this?

While I can't speak to the original intention, I understand the answer here to be "yes, it is deliberate" and I would be (generally) against including the range in the hashcode for nodes. 

In part, this is because not all nodes will have a range but my main concern is that adding/removing an import (or any other newline) would cause every node after it to have its hashcode recalculated. As a result, this would be a _substantive_ change in behaviour.

That said, I'm not completely opposed to code changes if a well-thought-through solution is put forward. I don't know what this would look like though!

Perhaps, as an immediate local solution @MorMundHS-MA , you could decorate any nodes you insert into a Map with a custom hashcode implementation? Or perhaps you could extend Map to take into account the range? Let me know if those suggestions don't make sense though!

**Comment by MysterAitch** [comment]
The javadoc for the `HashCodeVisitor` could probably use some editing/clarification to highlight whatever the outcome of this conversation is though!

e.g. adding a note to clarify "of all its **non-range** properties", or perhaps some other similar phrasing to show that the hashcode is calculated based on that node and its children (NOT its parents).


```java
/**
 * A visitor that calculates a deep hash code for a node by using the hash codes of all its properties,
 * and the hash codes of all its child nodes (by visiting those too.)
 */
public class HashCodeVisitor implements GenericVisitor<Integer, Void> {
```

https://github.com/javaparser/javaparser/blob/f20d6fdc306bb8ae135e9a9639ca1b04acecd976/javaparser-core/src/main/java/com/github/javaparser/ast/visitor/HashCodeVisitor.java#L31-L39

**Comment by MorMundHS-MA** [comment]
> Perhaps, as an immediate local solution @MorMundHS-MA , you could decorate any nodes you insert into a Map with a custom hashcode implementation?

@MysterAitch Yep, that's what I did. I had to wrap the nodes in another class anyway.

> In part, this is because not all nodes will have a range but my main concern is that adding/removing an import (or any other newline) would cause every node after it to have its hashcode recalculated. As a result, this would be a substantive change in behaviour.

Agreed. Calculating the hash code from mutable properties is bad design. And making all nodes immutable sounds impractical too. But, I guess this is technical a problem already.

> The javadoc for the HashCodeVisitor could probably use some editing/clarification to highlight whatever the outcome of this conversation is though!

The problem with documenting this is that no one calls hashCode or the HashCodeVisitor directly. Most use-cases will be maps or sets, which break silently. So while documenting it definitely won't hurt, it will still require some digging by the user. I'm personally curios how many people actually use maps/hashCodes with this project, because nobody has noticed the collision before, as far as I can tell. So my suggestion would be to make it obvious that nodes aren't hashable. Either by throwing an exception in hashCode or just hashing the type of the node so they always collide.

**Comment by matozoid** [comment]
Sorry I forgot about this: such a map (and set and list) is available here https://github.com/javaparser/javaparser/blob/master/javaparser-core/src/main/java/com/github/javaparser/utils/VisitorMap.java

**Comment by jlerbsc** [comment]
Do these responses answer your questions? If yes maybe can you close this issue.
